content/docs/user-guide/machine-learning-pipelines/defining-pipelines.md

```diff
@@ -0,0 +1,264 @@
+# Defining Pipelines
+
+DVC helps define your ML workflows so that anyone can reliably **reproduce**
+them later. This way you can ensure that steps are followed when necessary (and
+only when necessary).
+
+Specifically, pipelines are written as a set of `stages` in
+[`dvc.yaml` metafiles](#dvcyaml-metafiles). This _codification_ has the added
+benefit of bringing your pipelining process onto standard Git workflows
+(GitOps).
+
+## Stages
+
+A pipeline is a collection of data processing stages related to one another.
+Stages wrap around an executable shell command and specify its inputs and
+outputs (if any).
+
+Stage execution order is determined by defining outputs that feed into
+subsequent stage inputs, e.g. if an output of stage X is used as an input in
+stage Y, then DVC deduces that X should be run before Y. Technically, this is
+called a _dependency graph_ (specifically a DAG).
+
+<admon type="info">
+
+Note that the order of execution is entirely based on their DAG, and not on the
+order in which stages are found in `dvc.yaml`.
+
+</admon>
+
+[specification]: /doc/user-guide/project-structure/dvcyaml-files#stage-entries
+
+<details>
+
+### Avoiding unexpected behavior
+
+We don't want to tell anyone how to write their code or what programs to use!
+However, please be aware that in order to prevent unexpected results when DVC
+reproduces pipeline stages, the underlying code should ideally follow these
+rules:
+
+- Read/write exclusively from/to the specified <abbr>dependencies</abbr> and
+  <abbr>outputs</abbr> (including parameters files, metrics, and plots).
+
+- Completely rewrite outputs. Do not append or edit.
+
+- Stop reading and writing files when the `command` exits.
+
+Also, if your pipeline reproducibility goals include consistent output data, its
+code should be
+[deterministic](https://en.wikipedia.org/wiki/Deterministic_algorithm) (produce
+the same output for any given input): avoid code that increases
+[entropy](https://en.wikipedia.org/wiki/Software_entropy) (e.g. random numbers,
+time functions, hardware dependencies, etc.).
+
+</details>
+
+## Directed acyclic graph (DAG)
+
+DVC represents a pipeline internally as a graph where the nodes are stages and
+the edges are _directed_ dependencies (e.g. A before B). In order for DVC to
+execute the pipeline reliably, its topology should be _acyclic_ -- because
+executing cycles (e.g. A -> B -> C -> A ...) would continue indefinitely. [More
+about DAGs].
+
+Use `dvc dag` to visualize (or export) pipeline DAGs.
+
+[more about dags]: https://en.wikipedia.org/wiki/Directed_acyclic_graph
+
+## `dvc.yaml` metafiles
+
+There are two ways to define <abbr>stages</abbr>. The first and recommended
+method is by writing `dvc.yaml` files directly (`stages` list). The other one is
```

Comment:
Even for me, it's still painful to write `dvc.yaml` from scratch. Realistically, it's much easier to copy and paste to start from some template. 

Would it make sense to have some example `dvc.yaml` that includes common fields like `deps`, `params`, `outs`, `metrics`, `plots`? I see one example below but maybe it's too simple?

Reply:
> have some example dvc.yaml that includes common fields like deps, params, outs, metrics, plots?

A simple one with `deps` and `outs` is 2 paragraphs down (see [review app](https://dvc-org-iesahin-ug-pipe-qykfpz.herokuapp.com/doc/user-guide/machine-learning-pipelines/defining-pipelines#dvcyaml-metafiles)). We're not getting to params, metrics and plots yet.